cmd/rest-server/listener_unix.go

```diff
@@ -0,0 +1,45 @@
+// +build !windows
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"net"
+
+	"github.com/coreos/go-systemd/activation"
+)
+
+// findListener tries to find a listener via systemd socket activation. If that
+// fails, it tries to create a listener on addr.
+func findListener(addr string) (listener net.Listener, err error) {
+	// try systemd socket activation
+	listeners, err := activation.Listeners()
+	if err != nil {
+		panic(err)
+	}
+
+	switch len(listeners) {
+	case 0:
+		// no listeners found, listen manually
+		listener, err = net.Listen("tcp", addr)
+		if err != nil {
+			return nil, fmt.Errorf("listen on %v failed: %w", addr, err)
+		}
+
+		log.Printf("start server on %v", addr)
+		return listener, nil
+
+	case 1:
+		// one listener supplied by systemd, use that one
+		//
+		// for testing, run rest-server with systemd-socket-activate as follows:
+		//
+		//    systemd-socket-activate -l 8080 ./rest-server
+		log.Printf("systemd socket activation mode")
+		return listeners[0], nil
```

Comment:
I'm not sure who should do the check, but if the environmental flags don't match the passed in file descriptors, `listerners[0]` can be nil here and result in a panic later. Not a very nice to debug.

Reply:
Checking that `listeners[0]` is valid is probably a good idea before returning it. The expected behavior of `findListener` is that it either returns a listening socket or an error. Returning `nil` is thus wrong. Would you be willing to open a PR to check that `listeners[0] != nil`?

Reply:
I've opened https://github.com/coreos/go-systemd/issues/405 for now, let's see if they are willing to fix the condition on their end.